### PR TITLE
feat: add option to ignore urls for auto waiting

### PR DESCRIPTION
**Pull request by sebbi08**
Add the option to skip the XHRWaiter an the fetchWaiter based on an regex in the wdi5 config.

### CODE REVIEW

**Comment by sebbi08** [comment]
Could you hint me where in the Docs this should be documented ?

File: client-side-js/injectXHRPatch.cjs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+                return fnOriginalFakeOpen.apply(this, hooketoXHROpen.apply(this, arguments))
+            }
+
+            // Hook into XHR open for regular XHRs
+            const fnOriginalOpen = XMLHttpRequest.prototype.open
+            XMLHttpRequest.prototype.open = function () {
+                return fnOriginalOpen.apply(this, hooketoXHROpen.apply(this, arguments))
+            }
+
+            function hooketoXHROpen(method, url, async) {
```
Comment (Siolto) [inline client-side-js/injectXHRPatch.cjs:28]:
```suggestion
            function hooktoXHROpen(method, url, async) {
```

**Review by Siolto — CHANGES_REQUESTED** [review]
Type definition for new optional wdi5 property is missing [here](https://github.com/ui5-community/wdi5/blob/6421e8055ef2cb1789cb421a077ec26659837fe1/src/types/wdi5.types.ts#L26)

Documentation should be added somewhere [here](https://github.com/ui5-community/wdi5/blob/6421e8055ef2cb1789cb421a077ec26659837fe1/docs/configuration.md?plain=1#L24) I guess.

File: docs/configuration.md [hunk 2]
```diff
@@ -133,6 +134,12 @@ Number in milliseconds (default: `15000`) to wait for UI5-related operations wit
 Boolean setting to trigger injecting `wdi5` into both the shell and the app when used with the SAP Build Workzone, standard edition.  
 Recommended complement is to also [configure IAS Authentication](authentication?id=sap-cloud-identity-services-identity-authentication): as SAP Build requires its own Identity Provider (most likely provided by using an IAS tenant), you'll have to configure authentication against that as well in `wdi5`.
 
+### `autoWaitUrlIgnoreRegex`
+
+Ignore list in form of regex. Those will be used to ignore certain XHR/Fetch call from beeing waited for by the OPA5 Waiters. This can be used in combination with longpolling requests to continusly update your app.
```
Comment (vobu) [inline docs/configuration.md:139]:
```suggestion
Array of URLs (as strings), either relative or absolute. `RegEx` are supported.   
Querying the URLs will be excluded from the UI5 lifecycle sync. Meaning: no Test code will wait until querying the URLs resolve. 
Typical use case is "longpolling" requests that continuously update your app.
```

File: docs/configuration.md [hunk 3]
```diff
@@ -133,6 +134,12 @@ Number in milliseconds (default: `15000`) to wait for UI5-related operations wit
 Boolean setting to trigger injecting `wdi5` into both the shell and the app when used with the SAP Build Workzone, standard edition.  
 Recommended complement is to also [configure IAS Authentication](authentication?id=sap-cloud-identity-services-identity-authentication): as SAP Build requires its own Identity Provider (most likely provided by using an IAS tenant), you'll have to configure authentication against that as well in `wdi5`.
 
+### `autoWaitUrlIgnoreRegex`
```
Comment (vobu) [inline docs/configuration.md:137]: see my rename comment

File: examples/ui5-js-app/scripts/delayedMockServer.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
+const femiddleware = require("@sap-ux/ui5-middleware-fe-mockserver")
+
+module.exports = async function middleware(middlewareConfig) {
```
Comment (vobu) [inline examples/ui5-js-app/scripts/delayedMockServer.js:3]: this is so cool. well well done!!!

File: examples/ui5-js-app/ui5.yaml [hunk 5]
```diff
@@ -10,3 +10,19 @@ server:
       configuration:
         baseUri: "https://services.odata.org/V2"
         strictSSL: false
+    - name: delayed-mockserver
```
Comment (vobu) [inline examples/ui5-js-app/ui5.yaml:13]: again: cool idea, well done 😀

File: examples/ui5-js-app/webapp/Component.js [hunk 6]
```diff
@@ -27,6 +27,20 @@ sap.ui.define(
...
+
+                const url = new URL(location.href)
+                if (url.searchParams.get("isui5toolingTest")?.toLocaleLowerCase() === "true") {
+                    const startXHR = () => {
+                        this.getModel().read("/Categories", {
+                            success: startXHR
+                        })
+                    }
+                    startXHR()
+                    const startFetch = () => {
```
Comment (vobu) [inline examples/ui5-js-app/webapp/Component.js:39]: i keep my fingers crossed that the gh runners are equipped with enough memory for the CI runs 😁

**Review by vobu — COMMENTED** [review]
main thing is my naming suggestion (don't hate me, please)

File: src/lib/wdi5-bridge.ts [hunk 7]
```diff
@@ -135,6 +136,7 @@ export async function injectUI5(config: wdi5Config, browserInstance) {
     const version = await (browserInstance as WebdriverIO.Browser).getUI5Version()
     await checkUI5Version(version)
     await clientSide_injectTools(browserInstance) // helpers for wdi5 browser scope
+    await clientSide_injectXHRPatch(config, browserInstance)
```
[the remark this answers is not in the document]
Comment (sebbi08) [reply src/lib/wdi5-bridge.ts:139]: hi sadly not the clientSide_injectTools function add the compare version function to the browser and that is needed in clientSide_injectXHRPatch and checkUI5Version is actualy not async.

File: src/types/wdi5.types.ts [hunk 8]
```diff
@@ -53,6 +53,11 @@ export interface wdi5Config extends WebdriverIO.Config {
          * in a SAP Build Workzone Standard Edition (fka Launchpad) environment
          */
         btpWorkZoneEnablement?: boolean
+        /**
+         * Regex for XHR/Fetch requests to be ignored by the auto waiter
+         * Ideal for long polling as this would result in the waiter waiting forever
+         */
+        autoWaitUrlIgnoreRegex?: string[]
```
[the remark this answers is not in the document]
Comment (sebbi08) [reply src/types/wdi5.types.ts]: haha no hate here :D my naming skills are way to verbal. I am fine with that.

**Pull request merged**